src/hotspot/share/cds/filemap.cpp

```diff
@@ -1764,12 +1764,17 @@ size_t FileMapInfo::write_heap_regions(GrowableArray<MemRegion>* regions,
 
 void FileMapInfo::write_bytes(const void* buffer, size_t nbytes) {
   assert(_file_open, "must be");
-  ssize_t n = os::write(_fd, buffer, (unsigned int)nbytes);
-  if (n < 0 || (size_t)n != nbytes) {
-    // If the shared archive is corrupted, close it and remove it.
-    close();
-    remove(_full_path);
-    fail_stop("Unable to write to shared archive file.");
+  ssize_t size = (ssize_t)nbytes;
```

Comment:
I'm very confused as to how the underlying OS `write` can take in an unsigned count of bytes to write, yet the actual count of what was written is signed and so cannot reflect the true count for large values! Can we at least assert here that size is > 0 ?

Reply:
Can this be the rationale for it?
The return value can be `-1` so the type is `ssize_t`. And it is also large enough to hold the maximum bytes transfer of `write()`:
From the [write() man page](https://man7.org/linux/man-pages/man2/write.2.html#:~:text=of%20bytes%20written.-,On%20Linux%2C,-write()%20(and%20similar)) 
```
       On Linux, write() (and similar system calls) will transfer at
       most 0x7ffff000 (2,147,479,552) bytes, returning the number of
       bytes actually transferred.  (This is true on both 32-bit and
       64-bit systems.)

```

Reply:
`size` is checked to be `> 0` in loop's condition. Is assert for it still needed?

Reply:
With the new information an assert that nbytes <= SSIZE_MAX might still be appropriate. Let's see what others think.